### PR TITLE
Add Result::cloned{,_err} and Result::copied{,_err}

### PR DESCRIPTION
**Pull request by ksqsf**
This is a little nice addition to `Result`. 

1. I'm not sure how useful are `cloned_err` and `copied_err`, but for the sake of completeness they are here.
2. Naming is similar to `map`/`map_err`. I thought about naming `cloned` as `cloned_ok` and add another method called `cloned` that clones both Ok and Err, but `cloned_ok` should be more prevalent than `cloned_both`.

### CODE REVIEW

File: src/libcore/result.rs [hunk 1]
```diff
@@ -820,6 +820,166 @@ impl<T, E> Result<T, E> {
...
+    /// ```
+    /// #![feature(result_copied)]
+    /// let val = 12;
+    /// let x = Ok(&val);
+    /// assert_eq!(x, Ok(&12));
+    /// let copied = x.copied();
+    /// assert_eq!(copied, Ok(12));
+    /// ```
+    #[unstable(feature = "result_copied", reason = "newly added", issue = "63168")]
+    pub fn copied(self) -> Result<T, E> {
```
Comment (Centril) [inline src/libcore/result.rs]: My gut tells me that this should be called `copied_ok` so that there can be a `copied` that will copy both sides in one go.

Comment (ksqsf) [reply src/libcore/result.rs]: I'm not entirely sure about the usefulness. From my experience, I usually only need `copied_ok` and `cloned_ok`. But indeed having `copied`, `copied_ok`, and `copied_err` is more consistent. I've added them in the update.

**Comment by sfackler** [comment]
Is there any context in which cloned_err would be used?

**Comment by ksqsf** [comment]
@sfackler Actually I didn't think of a single use case. I'd be more than happy if we could remove `_err`, but as has been said, that's maybe just me. I'd suggest leaving it here in case someone finds it useful.

Introducing `cloned` & `copied` that work on both Ok and Err brings a lot of duplicate code, which is unwanted. (And I forgot to change the doc tests, oops.) I'm going to revert the last change.

**Comment by sfackler** [comment]
Yeah, let's remove the error variants.

Come to think of it, I'm having a hard time thinking of a context where I had a `Result<&T, E>` either. Out of curiosity, where does that come up for you?

**Comment by ksqsf** [comment]
My use case for `cloned`/`copied`: I have a cache around a database, and queries will return shared references `Result<&T, Err>` to data held by the cache. Whenever I want to make changes and write them back, I need to clone the contents. I did try `cloned` at the time but later found out only `Option` has got it.

**Comment by ProgrammaticNajel** [comment]
Ping from triage @sfackler any updates on this? Thanks.

**Comment by ksqsf** [comment]
ping @sfackler

**Comment by gagan0723** [comment]
Second ping from triage, @sfackler can you please review this.

Thanks

**Comment by Alexendoo** [comment]
Ping from triage, requesting a review @rust-lang/libs

**Comment by Dylan-DPC** [comment]
@ksqsf can you resolve the conflicts while we get either them or someone from the team to review it? thanks

**Comment by ksqsf** [comment]
@Dylan-DPC Hi, I just ran the tests locally (`./x.py test --doc src/libcore`) and there are no failing tests. Let me know if there's something I overlooked.

**Comment by alexcrichton** [comment]
@bors: r+

This looks good to me to add as unstable. I added some explicit questions to the tracking issue to make sure we get back to them by stabilization.

**Pull request merged**